### ISSUE TITLE
Query on confidence function used in the loss function

### ISSUE THREAD

**Issue by Aznoryusof** [benfred/implicit#329]
Hi Ben,

Thanks for the great library!

I am using the implementation of the ALS model for a project that I am working on, and have a question regarding the confidence function that was used as part of the loss function.

According to the original paper ("Collaborative Filtering for Implicit Feedback Datasets"), the following were the choices one could take:
1) c_ui = 1 + alpha * r_ui
2) c_ui = 1 + alpha * log(1 + r_ui/epsilon)

Could you share with me which option was used in the library's implementation and where I could locate that in the code please?

Thank you!

**Comment by ita9naiwa** [comment]
Sorry for the late response.

r_ui comes from the value of the matrix.

you may set log(1 + r_ui/epsilon) using like

```python
epsilon = some real value
matrix = some csr_matrix
log_matrix = log(1.0 + matrix/epsilon)
model.fit(log_matrix)
```

**Comment by Aznoryusof** [comment]
Thanks ita9naiwa for the clarifications!

Very helpful :)

**Comment by medoidai** [comment]
Hi @ita9naiwa:

When you say:

``
c_ui = 1 # for negative items
``

You actually mean this applied on the missing interactions and that we do not have to set it manually in our data and that the library handle it. Right? Our matrix is csr_matrix and for the missing interactions we have the value 0 (implicitly due to csr_matrix). So, the matrix handle these interactions with c_ui=1.

**Comment by medoidai** [comment]
Thank you, it was answered in another issue.
